### ISSUE TITLE
When XLA is used, bert nvprof shows no compute and communication overlap 

### ISSUE THREAD

**Issue by LiweiPeng** [NVIDIA/DeepLearningExamples#156]
When I trained the bert model with horovod and XLA, although XLA can significantly improve the perf, it decreased the scalability significantly. 

Our cluster has 16 nodes, 4 V100 32GB/node. The nodes are linked with 100GB Mellanox RoCE RDMA. There is no NVLINK/NVSwitch technology used.

When XLA is not used, the scalability from 1 to 16 GPU is 1.92x. When XLA is used, the scalability dropped to 1.8x from 1 to 16 GPUs.

Environment:
Framework: (TensorFlow, Keras, PyTorch, MXNet) Tensorflow
Framework version: 1.14
Horovod version: 0.16.4
MPI version: 3.1.1
CUDA version: 10.0
NCCL version: 2.4.7
Python version: 2.7.5
OS and version: CentOS 7.4
GCC version: 4.8.5

When XLA is not used, nvprof shows that there is good compute and communication (shown as 'mem' in the screenshot below) overlap, as expected.
![a](https://user-images.githubusercontent.com/8562078/62963590-25f4ab80-bdb6-11e9-9a59-27eb48f0c52a.png)

However, when XLA is used, nvprof shows that there is little compute and communication overlap.
![b](https://user-images.githubusercontent.com/8562078/62963674-51779600-bdb6-11e9-8f9b-22ec9c29abb1.png)

The problem is: What caused this no compute and communication overlap when XLA is used? Is this a bug or expected behavior for XLA?

**Comment by trentlo** [comment]
Thanks for reporting the issue.

It sounds like an XLA clustering issue because some parallelism is lost after clustering. If you can provide a reproducer, I can take a look of it to confirm my guess. And if it is what I guess, I am happy to provide a patch to fix it.

**Comment by LiweiPeng** [comment]
Thanks @trentlo . The way this can be reproduced: 
- use one node with like 4 GPUs. set up software environment according to my report.
- run nvidia BERT https://github.com/NVIDIA/DeepLearningExamples/tree/master/TensorFlow/LanguageModeling/BERT with --horovod --manual_fp16 --use-xla. 

I ran the above under nvprof when I did the nv profiling: nvprof --export-profile tf.%p.%h.nvprof python run_pretraining.py [options]

**Comment by LiweiPeng** [comment]
@trentlo Any update on this issue? How can I get your patch?

**Comment by trentlo** [comment]
> @trentlo Any update on this issue? How can I get your patch?
I am able to see the salability loss you described but I am still finding a way to confirm the source of the issue.

**Comment by LiweiPeng** [comment]
@trentlo Trent, Thanks for the quick update. Looking forward to using your solution.

**Comment by trentlo** [comment]
Just a quick update that the initial guess appears not correct. This seems not an XLA clustering issue.

Quick note:
If this is an XLA clustering issue, we should see a slowdown (compared to native TF) running XLA with the following flag `TF_XLA_FLAGS="--tf_xla_always_defer_compilation=true"`, which maintains the same clustering structures while all clusters are executed by the native TF (instead of XLA.) I, however, see the performance (of XLA with the flag) being on par with native TF.

That's been said, I will still be interested in understanding why such a slowdown by some detailed profiling. That will take a while though.

**Comment by LiweiPeng** [comment]
Thanks @trentlo  for the detailed analysis and update. Because if the problem is resolved, it can significantly improve GPU performance and scalability. It worths to do deep analysis.

**Comment by trentlo** [comment]
@LiweiPeng 

I found out that one major source of the slowdown, switch from 1 GPU to 2 GPUs, is because the grappler memory optimizer inserting some swaps (i.e., DataCopy from device to host and then from host to device). XLA loses more performance due to these swaps than the pure TF, because clustering can reduce the possible overlap (of these swaps and compute).

I did the experiments on 16GB DGX-1 with batch_size=3. I am not sure if you hit the same issue or not. You can try the below diff to turn off grappler memory optimization to see the performance impact. These swaps are purely overhead and gain nothing in my runs.

BTW, how do you calculate the scalability? I just noticed that your reported scalability does not look so bad (as the performance I am observing.) In my runs, XLA on 1GPU can speed up the BERT by ~55%, while only ~18% on multiple GPUs.

```
diff --git a/TensorFlow/LanguageModeling/BERT/run_pretraining.py b/TensorFlow/LanguageModeling/BERT/run_pretraining.py
index 450bded..8c6dcad 100644
--- a/TensorFlow/LanguageModeling/BERT/run_pretraining.py
+++ b/TensorFlow/LanguageModeling/BERT/run_pretraining.py
@@ -23,6 +23,7 @@ import time
 import modeling
 import optimization
 import tensorflow as tf
+from tensorflow.core.protobuf import rewriter_config_pb2
 
 flags = tf.flags
 
@@ -467,6 +468,10 @@ def main(_):
         raise ValueError("Input Files must be sharded")
   if FLAGS.use_xla: 
     config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1
+
+  # Turn off mem opt.
+  config.graph_options.rewrite_options.memory_optimization = rewriter_config_pb2.RewriterConfig.NO_MEM_OPT
+
   is_per_host = tf.contrib.tpu.InputPipelineConfig.PER_HOST_V2
   config = tf.ConfigProto()
   if FLAGS.horovod:

```

**Comment by LiweiPeng** [comment]
@trentlo Thanks for the awesome deep analysis and code. I'll test the option you provided.

For the XLA scalability, I measured the BERT throughput in examples/second in terms of number of GPUs. For example, the following graph is one with XLA plus mixed-precision. 
![1](https://user-images.githubusercontent.com/8562078/63625681-2dd1fe00-c5b5-11e9-95fa-32fd6f92c2d6.png)

Without XLA, the scalability is like 1.9x (not posted here)

**Comment by trentlo** [comment]
Thanks for the reply. You may try the provided diff to see if you hit the same issue.

However, by inspecting your numbers, you may not hit the same issue I'm observing. Please note that Horovod has some fixed overhead (~60ms in my measurement). So, it is nature that it impacts XLA more (percentage-wise) than the native TF according to the Amdahl's Law.

Please still let me know your test results though.

BTW, I also found that using `TF_XLA_FLAGS="--tf_xla_max_cluster_size=500"` gains me some performance. If you are interested in it, you may give it a try too.

**Comment by LiweiPeng** [comment]
Thanks Trent. I'll try both options and report test results.

**Comment by trentlo** [comment]
Somehow I just noticed your reply (perhaps because editing the comment does not trigger new notification?).

1.  Yes to your 3rd note. Clustering still has some issues due to loss of concurrency. To fix them, we need to make some structural change to Tensorflow, which may not be favorable now. Fortunately, the performance impact is not huge (besides the memory opt overhead, which you do not observe).
2. And yes to your 2nd note, one of my colleagues within NVIDIA also confirms that the Grappler memory opt is not triggered on 32GB GPUs.
3. I wrote up a case study document on this interesting BERT case. You can read it through the following link: https://docs.google.com/document/d/1q1UPN2CRRNBoUXM0zORT-cTG9m7ctQRB2xfUPhaK1Ek/edit?usp=sharing

Overall, yes, the XLA clustering will change the execution order and loses some concurrency although the performance impact is not huge (within 10%). And note that the Horovod will introduce some fixed overhead, which XLA cannot optimize, and we lose some speedup because of it.

**Comment by LiweiPeng** [comment]
@trentlo Really appreciate your detailed feedback. I'll study your bert xla case document in the above link.

**Comment by LiweiPeng** [comment]
Thanks @trentlo for the detailed analysis. You may close the issue now.

FYI, when I did the perf test above, my batch size is about 80% to the max batch size supported on the 32GB V100. I'll try to increase the batch size later to see the difference.
